src/drivers/uavcan/sensors/gnss.cpp

```diff
@@ -396,3 +427,73 @@ void UavcanGnssBridge::process_fixx(const uavcan::ReceivedDataStructure<FixType>
 
 	publish(msg.getSrcNodeID().get(), &report);
 }
+
+void UavcanGnssBridge::update()
+{
+	handleInjectDataTopic();
+}
+
+// Partially taken from src/drivers/gps/gps.cpp
+// This listens on the gps_inject_data uORB topic for RTCM data
+// sent from a GCS (usually over MAVLINK GPS_RTCM_DATA).
+// Forwarding this data to the UAVCAN bus enables DGPS/RTK GPS
+// to work.
+void UavcanGnssBridge::handleInjectDataTopic()
+{
+	bool updated = false;
+
+	// Limit maximum number of GPS injections to 6 since usually
+	// GPS injections should consist of 1-4 packets (GPS, Glonass, BeiDou, Galileo).
+	// Looking at 6 packets thus guarantees, that at least a full injection
+	// data set is evaluated.
+	const size_t max_num_injections = 6;
+	size_t num_injections = 0;
+
+	do {
+		num_injections++;
+		updated = _orb_inject_data_sub.updated();
+
+		if (updated) {
+			gps_inject_data_s msg;
+
+			if (_orb_inject_data_sub.copy(&msg)) {
+
+				/* Write the message to the gps device. Note that the message could be fragmented.
+				 * But as we don't write anywhere else to the device during operation, we don't
+				 * need to assemble the message first.
+				 */
+				injectData(msg.data, msg.len);
+			}
+		}
+	} while (updated && num_injections < max_num_injections);
+}
+
+bool UavcanGnssBridge::injectData(const uint8_t *data, size_t len)
+{
+	using uavcan::equipment::gnss::RTCMStream;
+
+	RTCMStream msg;
+	msg.protocol_id = RTCMStream::PROTOCOL_ID_RTCM3;
+
+	size_t capacity = msg.data.capacity();
+
+	if (len > capacity) {
+		return injectData(data, capacity) &&
+		       injectData(data + capacity, len - capacity);
```

Comment:
Can you avoid the recursion and expand this into a loop? Otherwise there's data-dependent stack size usage.

Reply:
Good point, I have rewritten this